client/netdb.c

```diff
@@ -64,41 +81,74 @@ static unsigned n_selected_hosts = 0;
 uint32_t *g_xdag_blocked_ips, *g_xdag_white_ips;
 int g_xdag_n_blocked_ips = 0, g_xdag_n_white_ips = 0;
 
+static void* check_alive_thread(void *host);
+
 static struct host *find_add_host(struct host *h)
 {
-	struct host *h0;
+	struct host *the_host;
 
 	pthread_mutex_lock(&host_mutex);
 
-	if ((h0 = (struct host *)ldus_rbtree_find(root, &h->node))) {
-		if (h->flags & HOST_SET) h0->flags = h->flags;
-	} else if (!(h->flags & HOST_NOT_ADD) && (h0 = malloc(sizeof(struct host)))) {
-		memcpy(h0, h, sizeof(struct host));
-		ldus_rbtree_insert(&root, &h0->node);
-		g_xdag_stats.nhosts++;
+	if ((the_host = (struct host *)ldus_rbtree_find(hosts_tree_root, &h->node))) {
+		if (h->flags & HOST_SET) {
+			the_host->flags = h->flags;
+		}
+		
+		if (!(the_host->flags & HOST_ACTIVE)) {
+			if (!(the_host->flags & HOST_INDB)) {
+				time_t now_time = time(0);
+				
+				if(now_time > (the_host->lasttime + 64)) {
+					pthread_t th;
+					int err = pthread_create(&th, 0, check_alive_thread, the_host);
+					if(err != 0) {
+						xdag_err("create check_alive_thread failed! %s", strerror(err));
+					}
+					
+					err = pthread_detach(th);
+					if(err != 0) {
+						xdag_err("detach check_alive_thread failed! %s", strerror(err));
+					}
+				}
+			}
+		}
+	} else if (!(h->flags & HOST_NOT_ADD) && (the_host = malloc(sizeof(struct host)))) {
+		memcpy(the_host, h, sizeof(struct host));
+		
+		ldus_rbtree_insert(&hosts_tree_root, &the_host->node);
 		
+		g_xdag_stats.nhosts++;
 		if (g_xdag_stats.nhosts > g_xdag_stats.total_nhosts) {
 			g_xdag_stats.total_nhosts = g_xdag_stats.nhosts;
 		}
-
+		
 		if (!(h->flags & HOST_INDB)) {
-			FILE *f = xdag_open_file(DATABASE, "a");
-			if (f) {
-				fprintf(f, "%u.%u.%u.%u:%u\n", h->ip & 0xff, h->ip >> 8 & 0xff, h->ip >> 16 & 0xff, h->ip >> 24 & 0xff, h->port);
-				xdag_close_file(f);
+			time_t now_time = time(0);
+			
+			if(now_time > (h->lasttime + 64)) {
+				pthread_t th;
+				int err = pthread_create(&th, 0, check_alive_thread, the_host);
+				if(err != 0) {
+					xdag_err("create check_alive_thread failed! %s", strerror(err));
+				}
+				
+				err = pthread_detach(th);
+				if(err != 0) {
+					xdag_err("detach check_alive_thread failed! %s", strerror(err));
+				}
 			}
 		}
 	}
 
 	if (h->flags & HOST_WHITE) {
 		if (g_xdag_n_white_ips < MAX_WHITE_IPS) {
-			g_xdag_white_ips[g_xdag_n_white_ips++] = h0->ip;
+			g_xdag_white_ips[g_xdag_n_white_ips++] = h->ip;
```

Comment:
You renamed h0 to the_host, but here you use variable 'h'

Reply:
h0 is malloc one, if not malloced it will crash, so use h.